File: Day-01-Arrays/CPP/283_Move_Zeroes.cpp
/*
===============================================================================
                            DSA Interview Playbook
===============================================================================

Problem:
283. Move Zeroes

Link:
https://leetcode.com/problems/move-zeroes/

Pattern:
Two Pointers

Difficulty:
Easy

Approach:
Maintain a pointer indicating the position where the next non-zero element
should be placed. Swap the current non-zero element with the element at the
pointer and increment the pointer.

Time Complexity:
O(n)

Space Complexity:
O(1)

===============================================================================
*/

class Solution {
public:
    void moveZeroes(vector<int>& nums) {

        int left = 0;

        for (int right = 0; right < nums.size(); right++) {

            if (nums[right] != 0) {
                swap(nums[left], nums[right]);
                left++;
            }
        }
    }
};

File: Day-01-Arrays/Python/283_Move_Zeroes.py
"""
Problem: 283. Move Zeroes

Pattern: Two Pointers

Time Complexity: O(n)
Space Complexity: O(1)
"""

class Solution:
    def moveZeroes(self, nums: list[int]) -> None:

        left = 0

        for right in range(len(nums)):

            if nums[right] != 0:
                nums[left], nums[right] = nums[right], nums[left]
                left += 1